### PR TITLE
Check match for status.ReadyReplicas with spec.Replicas for kapp to wait till deployment reconciliation 

### LINKED ISSUE #448: Deployment with strategy RECREATE doesn't wait for reconciliation during redeploy.
**What steps did you take:**
Create a deployment with `strategy: Recreate` and `replicas: 1`. Do the initial `kapp deploy`. Make some changes to deployment. Redeploy with `kapp deploy` again. This time, you will see that it didn't wait for reconciliation. 

For more information, check https://github.com/vmware-tanzu/carvel-kapp/discussions/434

### PR DESCRIPTION
**Pull request by kumaritanushree**
**Fixes:**  Kapp is exiting without waiting for the reconciliation and saying deployment succeeded. 
**Cause:** Kapp has check on status.UnavailableReplicas to verify the status of deployment which was not enough for some specific case ( if kapp read the yaml before  status.UnavailableReplicas got updated. Ref: https://github.com/vmware-tanzu/carvel-kapp/issues/448 )
**Solution:** Added check on status.ReadyReplicas with spec.Replicas to have double confirmation about the status of deployment.

**Acceptance Criteria:**
> With deploy strategy "RECREATE", KAPP should wait for reconciliation for the operations:
- [x]  Fresh deploy of an application
- [x]  Update in existing deployment (like change image)
- [x]  Increasing number of replicas (scaleUp)
- [x]  Decreasing number of replicas (scaleDown)
- [x]  Delete

> With no deploy strategy, KAPP should wait for reconciliation for the operations:
- [x]  Fresh deploy of an application
- [x]  Update in existing deployment (like change image)
- [x]  Increasing number of replicas (scaleUp)
- [x]  Decreasing number of replicas (scaleDown)
- [x]  Delete

### CODE REVIEW

**Review by 100mik — APPROVED** [review]
LGTM!

**Review by sethiyash — APPROVED** [review]
Looks Good.

File: pkg/kapp/resourcesmisc/apps_v1_deployment.go [hunk 1]
```diff
@@ -75,6 +75,10 @@ func (s AppsV1Deployment) IsDoneApplying() DoneApplyState {
 			"Waiting for %d unavailable replicas", dep.Status.UnavailableReplicas)}
 	}
 
+	if dep.Spec.Replicas != nil && dep.Status.ReadyReplicas < *dep.Spec.Replicas {
```
Comment (cppforlife) [inline pkg/kapp/resourcesmisc/apps_v1_deployment.go:78]: what happened to the nil check we were talking about?

**Comment by criztovyl** [comment]
I tested this branch using [my reproducer](https://github.com/criztovyl/kapp-448), seems to fix it, thx! :)

File: pkg/kapp/resourcesmisc/apps_v1_deployment.go (hunk 1 above)
Comment (100mik) [reply pkg/kapp/resourcesmisc/apps_v1_deployment.go:78]: I think the pointer is not derefrenced if the first condition among these two return false?
`dep.Spec.Replicas != nil && dep.Status.ReadyReplicas < *dep.Spec.Replicas`
I think it is safe.

**Comment by ChristianCiach** [comment]
I appreciate to see this issue fixed in such a short time. Awesome work (as always)!

But just a small question: I see the word "ready" a lot in this PR. Does that mean that I *have to* define readiness-probes for my deployments for this fix to work?

We usually don't use readiness-probes for many of our deployments, but I still want kapp to wait for the deployment to be successful. For example, if the issue described at https://github.com/vmware-tanzu/carvel-kapp/issues/448 appears, kapp doesn't even wait for the image to be pulled successfully. This means that currently (if `strategy: recreate`) we don't even see an error if a referenced image does not exist (at least most of the time, since the issue does not *always* appear).

TL;DR: The fix should not mandate the user to define readiness-probes. I am not fluent in golang, so I am not sure if this is currently the case with this PR.

**Comment by ChristianCiach** [comment]
I think I can answer my own question: The "Reproducer" at https://github.com/criztovyl/kapp-448 doesn't define readiness-probes and @criztovyl still confirms this to be fixed. Thank you so much, awesome people of carvel and criztovyl, too!

File: pkg/kapp/resourcesmisc/apps_v1_deployment.go (hunk 1 above)
Comment (kumaritanushree) [reply pkg/kapp/resourcesmisc/apps_v1_deployment.go:78]: @cppforlife  we were talking about the nil check of "dep.Spec.Replicas != nil" and as it is added with && operator it will safely exit if dep.Spec.Replicas is nil.
